golang/cosmos/app/upgrade.go

```diff
@@ -10,61 +10,37 @@ import (
 	upgradetypes "github.com/cosmos/cosmos-sdk/x/upgrade/types"
 )
 
-var upgradeNamesOfThisVersion = []string{
-	"UNRELEASED_BASIC", // no-frills
-	"UNRELEASED_A3P_INTEGRATION",
-	"UNRELEASED_main",
-	"UNRELEASED_devnet",
-	"UNRELEASED_REAPPLY",
+// upgradeNamesOfThisVersion documents the current upgrade names and whether
+// each is "primary" (used to trigger store migrations, which can only run
+// once). An actual release should have exactly one primary upgrade name and any
+// number of non-primary upgrade names (each of the latter being associated with
+// a release candidate iteration after the first RC and used only for
+// pre-production chains), but for testing purposes, the master branch has
+// multiple primary upgrade names.
+var upgradeNamesOfThisVersion = map[string]bool{
+	"UNRELEASED_BASIC":           true,
+	"UNRELEASED_A3P_INTEGRATION": true,
+	"UNRELEASED_main":            true,
+	"UNRELEASED_devnet":          true,
+	"UNRELEASED_REAPPLY":         false,
 }
 
-// isUpgradeNameOfThisVersion returns whether the provided plan name is a
-// known upgrade name of this software version
-func isUpgradeNameOfThisVersion(name string) bool {
-	for _, upgradeName := range upgradeNamesOfThisVersion {
-		if upgradeName == name {
-			return true
-		}
-	}
-	return false
-}
-
-// validUpgradeName is an identity function that asserts the provided name
-// is an upgrade name of this software version. It can be used as a sort of
-// dynamic enum check.
-func validUpgradeName(name string) string {
```

Comment:
While no longer in use right now, this helper was extremely useful when we had to generate different core proposal configs depending on the upgrade name (different price feed depending on the network). Let keep it around.

Maybe restore the check that `name` is a valid upgrade name in `isPrimaryUpgradeName` ?

Reply:
`isPrimaryUpgradeName` already includes that check... if the provided name is not a valid upgrade name, then `isPrimary, found := upgradeNamesOfThisVersion[name]` will set both `isPrimary` and `found` to false, the latter resulting in a panic two lines down. And any non-panicky validity check could be trivially performed against `upgradeNamesOfThisVersion` in exactly the same way.

Reply:
What I'm asking here is to keep `validUpgradeName` around as it was useful for a switch case that we needed when conditionally generating core proposals. I figured calling it from `isPrimaryUpgradeName` was an easy way to keep it in use to satisfy golint checks.

Reply:
Yeah, for such a release I'd probably just generalize `upgradeNamesOfThisVersion` from `map[string]bool` to `map[string][]string` (with each value containing the correct oracle addresses). And if we want that kind of generalization to be obvious, I can update this PR from `map[string]bool` to `map[string]*upgradeMetaData` (where the basic upgradeMetaData is just `struct{ isPrimary bool }`).

Reply:
what is the concern with keeping part of the customization related to upgrade-name imperative instead of through configuration? In JS a `switch` case is perfectly acceptable to do this. The problem is that in golang we don't have types to help us check a typo wasn't made.

Reply:
It requires the upgrade names to be retyped and muddies their source of truth, creating another opportunity for error in release branch initialization (which in fact I did miss in #10088). And it's not actually imperative anyway, since that link demonstrates that there was in fact a static mapping from upgrade name to data that was serialized into a JSON template. Basically, it's unnecessary complexity.

Reply:
In that particular case there was only a config change indeed. I would like a flexible option that allows conditionally including some core proposals depending on the upgrade name. What do you suggest is the most robust approach?